src/test/java/org/broadinstitute/hellbender/utils/spark/SparkUtilsShardUnitTest.java

```diff
@@ -0,0 +1,355 @@
+                        new SparkUtils.PartitionLocatable<>(2, new SimpleInterval("2", 12, 50))
+                ));
+    }
+
+    @Test
+    public void testReadsPerShard() throws IOException {
+        //                      1                   2
+        //    1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7
+        //   [-----]
+        //           [-----]
+        //               [-----]
+        // ---------------------------------------------------------
+        //               [-----]
+        //               [-----]
+        //               [-----]
+        // ---------------------------------------------------------
+        //               [-----]
+        //                       [-----]
+        //                         [-----]
+        // ---------------------------------------------------------
+        //                                   [-----]
+        //                                           [-----]
+        //                                                   [-----]
+        // ---------------------------------------------------------
+        //
+        //     [-----]                                                 <- intervals
+        //                 [---------]
+        //                       [-----------------------]
+        //
+        //                      1                   2
+        //    1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7
+
+        // maxEndPartitionIndexes: [2, 1, 3, 3]
+
+        List<TestRead> reads = ImmutableList.of(
+                new TestRead(1, 3), new TestRead(5, 7), new TestRead(7, 9),
+                new TestRead(7, 9), new TestRead(7, 9), new TestRead(7, 9),
+                new TestRead(7, 9), new TestRead(11, 13), new TestRead(12, 14),
+                new TestRead(17, 19), new TestRead(21, 23), new TestRead(25, 27)
+        );
+
+        List<SimpleInterval> intervals = ImmutableList.of(
+                new SimpleInterval("1", 2, 4),
+                new SimpleInterval("1", 8, 12),
+                new SimpleInterval("1", 11, 22));
+
+        Iterator<Tuple2<SimpleInterval, Iterable<TestRead>>> it = SparkUtils.readsPerShard(reads.iterator(), intervals.iterator(), sequenceDictionary);
+        assertTrue(it.hasNext());
+        Tuple2<SimpleInterval, Iterable<TestRead>> next = it.next();
+        assertEquals(next._1(), intervals.get(0));
+        assertEquals(next._2(), ImmutableList.of(reads.get(0)));
+
+        assertTrue(it.hasNext());
+        next = it.next();
+        assertEquals(next._1(), intervals.get(1));
+        assertEquals(next._2(), ImmutableList.of(reads.get(2), reads.get(3), reads.get(4), reads.get(5), reads.get(6), reads.get(7), reads.get(8)));
+
+        assertTrue(it.hasNext());
+        next = it.next();
+        assertEquals(next._1(), intervals.get(2));
+        assertEquals(next._2(), ImmutableList.of(reads.get(7), reads.get(8), reads.get(9), reads.get(10)));
+
+        assertFalse(it.hasNext());
+    }
```

Comment:
The logic is heavily exercised by the integration tests for the various tools that use this feature (there are more to come after this PR), so the small-scale unit tests that cover certain edge cases, plus the integration tests, should be sufficient.